libkbfs/root_metadata_test.go

```diff
@@ -535,7 +535,186 @@ func TestRootMetadataFinalIsFinal(t *testing.T) {
 	require.NoError(t, err)
 
 	rmd.SetFinalBit()
-	_, err = rmd.MakeSuccessor(nil, fakeMdID(1), true)
+	_, err = rmd.MakeSuccessor(context.Background(), nil, fakeMdID(1), true)
 	_, isFinalError := err.(MetadataIsFinalError)
 	require.Equal(t, isFinalError, true)
 }
+
+func getAllUsersKeysForTest(
+	t *testing.T, config Config, rmd *RootMetadata, un string) []kbfscrypto.TLFCryptKey {
+	var keys []kbfscrypto.TLFCryptKey
+	for i := 1; i <= int(rmd.LatestKeyGeneration()); i++ {
+		key, err := config.KeyManager().(*KeyManagerStandard).getTLFCryptKeyUsingCurrentDevice(
+			context.Background(), rmd, KeyGen(i), true)
+		require.NoError(t, err)
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+// We always want misses for the tests below.
+type dummyNoKeyCache struct {
+}
+
+func (kc *dummyNoKeyCache) GetTLFCryptKey(_ tlf.ID, _ KeyGen) (kbfscrypto.TLFCryptKey, error) {
+	return kbfscrypto.TLFCryptKey{}, KeyCacheMissError{}
+}
+
+func (kc *dummyNoKeyCache) PutTLFCryptKey(_ tlf.ID, _ KeyGen, _ kbfscrypto.TLFCryptKey) error {
+	return nil
+}
+
+// Test upconversion from MDv2 to MDv3 for a private folder.
+func TestRootMetadataUpconversionPrivate(t *testing.T) {
+	config := MakeTestConfigOrBust(t, "alice", "bob", "charlie")
+	config.SetKeyCache(&dummyNoKeyCache{})
+	defer config.Shutdown()
+
+	tlfID := tlf.FakeID(1, false)
+	h := parseTlfHandleOrBust(t, config, "alice,alice@twitter#bob,charlie@twitter,eve@reddit", false)
+	rmd, err := makeInitialRootMetadata(InitialExtraMetadataVer, tlfID, h)
+	require.NoError(t, err)
+	require.Equal(t, rmd.LatestKeyGeneration(), KeyGen(0))
+	require.Equal(t, rmd.Revision(), MetadataRevision(1))
+	require.Equal(t, rmd.Version(), InitialExtraMetadataVer)
+
+	// set some dummy numbers
+	diskUsage, refBytes, unrefBytes := uint64(12345), uint64(4321), uint64(1234)
+	rmd.SetDiskUsage(diskUsage)
+	rmd.SetRefBytes(refBytes)
+	rmd.SetUnrefBytes(unrefBytes)
+
+	// key it once
+	done, _, err := config.KeyManager().Rekey(context.Background(), rmd, false)
+	require.NoError(t, err)
+	require.True(t, done)
+	require.Equal(t, rmd.LatestKeyGeneration(), KeyGen(1))
+	require.Equal(t, rmd.Revision(), MetadataRevision(1))
+	require.Equal(t, rmd.Version(), InitialExtraMetadataVer)
+
+	// prove charlie
+	config.KeybaseService().(*KeybaseDaemonLocal).addNewAssertionForTestOrBust(
+		"charlie", "charlie@twitter")
+
+	// rekey it
+	done, _, err = config.KeyManager().Rekey(context.Background(), rmd, false)
+	require.NoError(t, err)
+	require.True(t, done)
+	require.Equal(t, rmd.LatestKeyGeneration(), KeyGen(1))
+	require.Equal(t, rmd.Revision(), MetadataRevision(1))
+	require.Equal(t, rmd.Version(), InitialExtraMetadataVer)
```

Comment:
Should check to make sure that there's a new TLFEphemeralPublicKey in the reader key list.